File: backend/src/main/java/com/example/database/services/ReportService.java
package com.example.database.services;

import com.example.database.models.Report;
import com.example.database.models.ReportType;
import com.example.database.models.User;
import com.example.database.repository.ReportRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Optional;

/**
 * Serwis obsługujący operacje na encji {@link Report}.
 *
 * Klasa zawiera logikę biznesową związaną z zarządzaniem raportami,
 * wykorzystując {@link ReportRepository} do komunikacji z bazą danych.
 *
 * @author YourName
 * @version 1.0.0
 * @since 1.0.0
 */
@Service
@Transactional
public class ReportService {

    private final ReportRepository reportRepository;

    @Autowired
    public ReportService(ReportRepository reportRepository) {
        this.reportRepository = reportRepository;
    }

    /**
     * Pobiera raporty po typie.
     *
     * @param type Typ raportu
     * @return Lista raportów danego typu
     */
    public List<Report> getReportsByType(ReportType type) {
        return reportRepository.findByType(type);
    }

    /**
     * Pobiera raporty utworzone przez danego użytkownika.
     *
     * @param user Użytkownik
     * @return Lista raportów utworzonych przez użytkownika
     */
    public List<Report> getReportsByCreatedBy(User user) {
        return reportRepository.findByCreatedBy(user);
    }

    /**
     * Zapisuje raport.
     *
     * @param report Raport do zapisania
     * @return Zapisany raport
     */
    public Report saveReport(Report report) {
        return reportRepository.save(report);
    }

    /**
     * Usuwa raport.
     *
     * @param report Raport do usunięcia
     */
    public void deleteReport(Report report) {
        reportRepository.delete(report);
    }

    /**
     * Pobiera raport po identyfikatorze.
     *
     * @param id Identyfikator raportu
     * @return Optional zawierający raport, jeśli istnieje
     */
    public Optional<Report> getReportById(Long id) {
        return reportRepository.findById(id);
    }

    /**
     * Pobiera wszystkie raporty.
     *
     * @return Lista wszystkich raportów
     */
    public List<Report> getAllReports() {
        return reportRepository.findAll();
    }
}

File: frontend/src/services/reportService.js
// src/services/reportService.js
import apiService from './apiService';

const reportService = {
    // Pobieranie wszystkich raportów
    async getAllReports() {
        return await apiService.get('/database/reports');
    },

    // Pobieranie raportu po ID
    async getReportById(reportId) {
        return await apiService.get(`/database/reports/${reportId}`);
    },

    // Tworzenie raportu
    async createReport(reportData) {
        return await apiService.post('/database/reports', reportData);
    },

    // Usuwanie raportu
    async deleteReport(reportId) {
        return await apiService.delete(`/database/reports/${reportId}`);
    },

    // Pobieranie raportów określonego typu
    async getReportsByType(typeId) {
        return await apiService.get(`/database/reports/type/${typeId}`);
    },

    // Pobieranie raportów utworzonych przez określonego użytkownika
    async getReportsByUser(userId) {
        return await apiService.get(`/database/reports/user/${userId}`);
    },

    // Aktualizacja raportu
    async updateReport(reportId, reportData) {
        return await apiService.put(`/database/reports/${reportId}`, reportData);
    },

    // Pobieranie wszystkich typów raportów
    async getAllReportTypes() {
        return await apiService.get('/database/report-types');
    },

    // Pobieranie typu raportu po ID
    async getReportTypeById(typeId) {
        return await apiService.get(`/database/report-types/${typeId}`);
    }
};

export default reportService;